### ISSUE TITLE
support cross-compilation

### ISSUE THREAD

**Issue by stsp** [alexfru/SmallerC#55]
I tried to build smallerc for termux,
and quickly ran into troubles:
https://github.com/termux/termux-packages/pull/22569
It appears impossible to execute the
target smallerc during build. I suppose
the right solution is to generate the
bootstrap smallerc and target smallerc.
What do you think?
Would it be feasible to support cross-compilation
scenarios?

**Comment by stsp** [comment]
You don't need to know much details
about Termux. Lets just consider it a
generic cross-compile environment.
The problem is very simple:
Your makefile executes `./smlrcc` during
build. This is not possible when cross-compiling,
simply because `./smlrcc` has an incompatible
arch. I worked around the problem with
this simple trick:
https://github.com/termux/termux-packages/pull/22569/commits/e01f4515e526bb0da32240a25182dbf9807c4111#diff-521d3945b54ec864d39d6398b60b2f3706de316642812f04fa48e34356e753d7

Its good for termux, but probably not
for upstreaming here. The proper
solution is up to you to evaluate.
IMO you need a 2-staged build,
having a "bootstrap" smallerc built
first. This is the natively-build one.
Then you can build the target one.

Let me know if more details are needed.

**Comment by stsp** [comment]
Basically you need to build the
bootstrap smallerc with `$(CC_FOR_BUILD)`
and the target smallerc with `$(CC)`.
If `$(CC_FOR_BUILD)` is not set then
we are not cross-compiling, so 1 stage
is enough.
(this is of course a very simplified scheme,
but it should work)

**Comment by stsp** [comment]
With autoconf, the cross-compile mode
is entered if `--host` is specified and
configure can't natively run the built
executables. But you don't have autoconf,
so I think just checking `$(CC_FOR_BUILD)`
would save the day.

**Comment by rofl0r** [comment]
> `$(CC_FOR_BUILD)`

also (and probably better) known as `$(HOSTCC)` outside of autoconf's confusing nomenclature.

**Comment by stsp** [comment]
> also (and probably better) known as $(HOSTCC) outside of autoconf's confusing nomenclature.

Termux and Builroot also use CC_FOR_BUILD.
Since these are major players, I suppose
its too late to figth the autoconf traditions. :)

**Comment by rofl0r** [comment]
the linux kernel which uses HOSTCC, is also a major player :)

**Comment by stsp** [comment]
I suspect the difference is in the
intended use. CC_FOR_BUILD is
probably common for cross-compile
environments, so when you write
the recipe for one, you'll likely get
CC_FOR_BUILD.
OTOH kernel is not a cross-compiling
environment, so it uses HOSTCC for
a very small amount of internal tools that
are needs to be built before the main build
process starts.

**Comment by rofl0r** [comment]
in the end, the exact name used isn't that important, as long as it's possible to override the CC used for host dependencies, and it's documented somewhere. though i've personally been using HOSTCC for all of my projects since it's more widely used in Makefile-based build systems without autoconf, and its name makes it clearer what it does. namely to build things that need to run on the host.

**Comment by stsp** [comment]
Ah, actually I am starting to understand
that you disagree not with the variable
name, but with the autoconf terminology
all together?
Because --host to autoconf means the
target machine, and --build means build
machine.

And in kernel:
```
CROSS = #/usr/i386-linux-uclibc/usr/bin/i386-uclibc-
CC = $(CROSS)gcc
LD = $(CROSS)gcc
AR = $(CROSS)ar
STRIP = $(CROSS)strip
RANLIB = $(CROSS)ranlib
HOSTCC = gcc
```

So indeed it appears not only the
name, but the entire notion of "host"
is redefined.
Ok, yep, then I see the problem and
I will no longer claim that "any name
will suit" - its really a semantical difference.

**Comment by stsp** [comment]
So I guess autoconf was written
solely for gcc. :) Otherwise the one
has troubles justifying --host being
target, and --target being the
cross-compiler target...

**Comment by rofl0r** [comment]
> you disagree not with the variable name, but with the autoconf terminology all together?

right, the autoconf terminology is pretty confusing, and apparently not only to me, but also to others whenever the topic comes up... maybe that's why linus used his own terminology, which has since been adopted by 1000s of projects...

**Comment by stsp** [comment]
> comes up... maybe that's why linus used his own terminology, which has since been adopted by 1000s of projects...

In this case, one possibility is
to do something like this:
```
ifneq ($(CC_FOR_BUILD),)
  we_are_cross_compiling := 1
  HOSTCC := $(CC_FOR_BUILD)
else
  # just use CC in native-compile mode and skip bootstrap stage, so HOSTCC is unused
endif
```

You need such an "adapter" for the
cross-compiling environments.

**Comment by rofl0r** [comment]
> You need such an "adapter" for the cross-compiling environments.

why's that ? you can just `make HOSTCC=gcc CC=arm-linux-gcc`

and the makefile look like

```make
CC=gcc
HOSTCC=gcc
```
the makefile of my project netbsd-curses could be used as a template, since it has stuff that needs to be built twice if CC != HOSTCC.

**Comment by stsp** [comment]
> why's that ?

Because I don't think this:

> to be built twice if CC != HOSTCC.

is reliable.
Who knows if some idiot sets CC=gcc
and HOSTCC=clang? By your logic this
will force a cross-compile mode.
I agree that its fairly theoretical, and
in principle something like this is possible
too. But I'd rather check some var explicitly.

**Comment by rofl0r** [comment]
> Who knows if some idiot sets CC=gcc and HOSTCC=clang?

that would work too, just take a bit longer, as some stuff would be compiled twice. luckily we're talking C build times, not C++ ones, or even worse, rust :)

**Comment by stsp** [comment]
> that would work too, just take a bit longer,

If only that then probably (?) fine.
In larger projects, messing up cross-compile
mode can eg skip tests (you can't run
cross-compiled tests on a build system).

**Comment by alexfru** [comment]
> Your makefile executes ./smlrcc during build. This is not possible when cross-compiling, simply because ./smlrcc has an incompatible arch.

Why is it incompatible? You're building smlrcc for x86 and trying to run that on ARM instead? Just build it for ARM and run on ARM. I'm pretty sure that'll just work (unless your ARM is big-endian). I think there're NASM binaries for ARM too.

**Comment by alexfru** [comment]
If cross-compiling is so convoluted, maybe you could avoid it.
Or prebuild stuff.
Or just use one of the binary releases for DOS/Windows. I create those once in a while. Smlrcc will run OK under DOSBox.

**Comment by stsp** [comment]
In case of smallerc, its not difficult
at all. I spent many month cross-compiling
gcc and ended up with 3-staged compilation
IIRC. :) But for smallerc, you have 2 options:
- support 2-staged buld yourself, which basically
  means build 2 copy of smallerc in different dirs
  using different CCs (CC/HOSTCC or CC_FOR_BUILD)
- or you can just apply [this patch](https://github.com/termux/termux-packages/blob/e01f4515e526bb0da32240a25182dbf9807c4111/packages/smallerc/makefile.patch)

In the second case you let the external
scripts to build smallerc twice. At the end
it may be even better: I do not purge the
host build on Termux. Instead I install it
into /opt, so that packages that depend
on smallerc, can also be host-built. So
maybe just applying the patch, is optimal.
Of course you can do nothing at all, in
which case the patch will needs to be
carried by termux. Since the patch is
that small, this doesn't hurt at all. But
I don't see the reason why it can't be
applied.

**Comment by alexfru** [comment]
Is that patch really sufficient? You don't have any problem invoking `nasm` from the compiler while building SmallerC's standard C libraries?

**Comment by stsp** [comment]
> Is that patch really sufficient?

Passes CI and allows to build dependent
packages. So definitely.

> You don't have any problem invoking nasm from the compiler while building SmallerC's standard C libraries?

Termux CI installs host nasm from
ubuntu. Whatever is available in
ubuntu, makes no problems on
Termux CI. Smallerc is also available
in ubuntu, but only since Oracular,
whereas Termux uses Noble.

Also note that nasm path never
needs to be hard-coded. If smallerc
calls it as just "nasm" its perfectly
fine: cross-compile scripts can juggle
PATH variable to solve the alike problems.
What really problematic is `./smlrcc`.
That needs to be un-hardcoded.

**Comment by alexfru** [comment]
Make a pull request with that patch then and I'll merge it in.

**Comment by alexfru** [comment]
Addressed in #56.
